#Обязательное задание 4 - анализ рынка видеоигр

В качестве ответа пришлите файл .ipynb в формате задание - решение - комментарий/вывод (если необходимо)

Датасет содержит информацию о видеоиграх с 1980 по 2016 год.

##Задача 1. Загрузка данных
Загрузите файл, выведите первые 10 строк, типы данных и количество пропусков в каждом столбце.

In [69]:
import pandas as pd

In [70]:
df = pd.read_csv('games.csv')
df.head(10)

,Name,Platform,Year_of_Release,Genre,NA_sales,EU_sales,JP_sales,Other_sales,Critic_Score,User_Score,Rating
0,Wii Sports,Wii,2006.0,Sports,41.36,28.96,3.77,8.45,76.0,8,E
1,Super Mario Bros.,NES,1985.0,Platform,29.08,3.58,6.81,0.77,NaN,NaN,NaN
2,Mario Kart Wii,Wii,2008.0,Racing,15.68,12.76,3.79,3.29,82.0,8.3,E
3,Wii Sports Resort,Wii,2009.0,Sports,15.61,10.93,3.28,2.95,80.0,8,E
4,Pokemon Red/Pokemon Blue,GB,1996.0,Role-Playing,11.27,8.89,10.22,1.00,NaN,NaN,NaN
5,Tetris,GB,1989.0,Puzzle,23.20,2.26,4.22,0.58,NaN,NaN,NaN
6,New Super Mario Bros.,DS,2006.0,Platform,11.28,9.14,6.50,2.88,89.0,8.5,E
7,Wii Play,Wii,2006.0,Misc,13.96,9.18,2.93,2.84,58.0,6.6,E
8,New Super Mario Bros. Wii,Wii,2009.0,Platform,14.44,6.94,4.70,2.24,87.0,8.4,E
9,Duck Hunt,NES,1984.0,Shooter,26.93,0.63,0.28,0.47,NaN,NaN,NaN


In [71]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16715 entries, 0 to 16714
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16713 non-null  object 
 1   Platform         16715 non-null  object 
 2   Year_of_Release  16446 non-null  float64
 3   Genre            16713 non-null  object 
 4   NA_sales         16715 non-null  float64
 5   EU_sales         16715 non-null  float64
 6   JP_sales         16715 non-null  float64
 7   Other_sales      16715 non-null  float64
 8   Critic_Score     8137 non-null   float64
 9   User_Score       10014 non-null  object 
 10  Rating           9949 non-null   object 
dtypes: float64(6), object(5)
memory usage: 1.4+ MB


In [72]:
df.isna().sum()

,0
Name,2
Platform,0
Year_of_Release,269
Genre,2
NA_sales,0
EU_sales,0
JP_sales,0
Other_sales,0
Critic_Score,8578
User_Score,6701


##Ответ:
 Были выведены первые  10 строк всей таблицы, информация о количестве строк и столбцов, типы данных (float64(6), object(5)), количество пропущенных строк по всем стобцам. В столбце "name" не заполнены 2 строки,  в столбце "Year_of_Release" - 269 строк, в столбце "Genre" - 2 строки, "Critic_Score" - 8578 строк, в столбце "User_Score" - 6701 строк, в стобце "Rating" - 6766 строк не заполнено.

## Задача 2. Очистка данных
Приведите названия столбцов к нижнему регистру.
В user_score замените 'tbd' на NaN и объясните, что оно означает.
Приведите user_score к числовому типу, year_of_release — к целочисленному. (подсказка: для замены поможет метод .replace, для создания NaN - библиотека import numpy as np и np.nan, для преобразования к числовому типу используйте типы данных с поддержкой NaN - Int64, Float32 и тд)

In [73]:
df.columns = df.columns.str.lower()
df.head()

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating
0,Wii Sports,Wii,2006.0,Sports,41.36,28.96,3.77,8.45,76.0,8,E
1,Super Mario Bros.,NES,1985.0,Platform,29.08,3.58,6.81,0.77,NaN,NaN,NaN
2,Mario Kart Wii,Wii,2008.0,Racing,15.68,12.76,3.79,3.29,82.0,8.3,E
3,Wii Sports Resort,Wii,2009.0,Sports,15.61,10.93,3.28,2.95,80.0,8,E
4,Pokemon Red/Pokemon Blue,GB,1996.0,Role-Playing,11.27,8.89,10.22,1.00,NaN,NaN,NaN


In [77]:
import numpy as np

df['user_score'] = df['user_score'].replace('tbd', np.nan)

In [76]:
df['user_score'] = df['user_score'].astype(np.float32)
print(df['user_score'].dtype)
print(df['user_score'].isna().sum())

float32
9125


In [78]:
df['year_of_release'] = df['year_of_release'].astype('Int64')
print(df['year_of_release'].dtype)
print(df['year_of_release'].isna().sum())

Int64
269


In [79]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16715 entries, 0 to 16714
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16713 non-null  object 
 1   platform         16715 non-null  object 
 2   year_of_release  16446 non-null  Int64  
 3   genre            16713 non-null  object 
 4   na_sales         16715 non-null  float64
 5   eu_sales         16715 non-null  float64
 6   jp_sales         16715 non-null  float64
 7   other_sales      16715 non-null  float64
 8   critic_score     8137 non-null   float64
 9   user_score       7590 non-null   float32
 10  rating           9949 non-null   object 
dtypes: Int64(1), float32(1), float64(5), object(4)
memory usage: 1.4+ MB


##Вывод:
1. Название столбцов приведено к нижнему регистру.
2. В столбце 'user_score' с помощью метода  .replace заменили 'tbd' на NaN, так как 'tbd'-это текстовая отметка и она мешает анализу и подсчету среднестатистических данных.
3. В столбце 'user_score' тип данных был преобразован в числовой (или дробный) 'float32 ', в столбце 'year_of_release' тип данных был преобразован в целочисленный 'Int64'.


##Задача 3. Суммарные продажи
Добавьте столбец total_sales — сумму продаж по всем регионам.
Выведите топ-10 игр по мировым продажам.
Найдите игру с наибольшими продажами отдельно в NA, EU и JP.

In [80]:
df['total_sales'] = df['na_sales'] + df['eu_sales'] + df['jp_sales'] + df['other_sales']
df['total_sales']

,total_sales
0,82.54
1,40.24
2,35.52
3,32.77
4,31.38
...,...
16710,0.01
16711,0.01
16712,0.01
16713,0.01


In [81]:
df.head(5)

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating,total_sales
0,Wii Sports,Wii,2006,Sports,41.36,28.96,3.77,8.45,76.0,8.0,E,82.54
1,Super Mario Bros.,NES,1985,Platform,29.08,3.58,6.81,0.77,NaN,NaN,NaN,40.24
2,Mario Kart Wii,Wii,2008,Racing,15.68,12.76,3.79,3.29,82.0,8.3,E,35.52
3,Wii Sports Resort,Wii,2009,Sports,15.61,10.93,3.28,2.95,80.0,8.0,E,32.77
4,Pokemon Red/Pokemon Blue,GB,1996,Role-Playing,11.27,8.89,10.22,1.00,NaN,NaN,NaN,31.38


Столбец "total_sales" был добавлен в общую таблицу.

In [82]:
top_10_games= df.sort_values('total_sales', ascending=False)

top_10_games[['name', 'total_sales']].head(10)

,name,total_sales
0,Wii Sports,82.54
1,Super Mario Bros.,40.24
2,Mario Kart Wii,35.52
3,Wii Sports Resort,32.77
4,Pokemon Red/Pokemon Blue,31.38
5,Tetris,30.26
6,New Super Mario Bros.,29.80
7,Wii Play,28.91
8,New Super Mario Bros. Wii,28.32
9,Duck Hunt,28.31


В таблице указаны топ-10 игр по мировым продажам (название игры - суммарные продажи по миру)

In [83]:
games_region_na = df.sort_values('na_sales', ascending=False)
games_region_na[['name', 'na_sales']].head(1)

,name,na_sales
0,Wii Sports,41.36


Игра "Wii Sports" занимает лидирующую позицию по продажам в Северной Америке (NA)

In [84]:
games_region_eu = df.sort_values(by='eu_sales', ascending=False)
games_region_eu[['name', 'eu_sales']].head(1)

,name,eu_sales
0,Wii Sports,28.96


Игра "Wii Sports" занимает лидирующую позицию по продажам в Европе (EU)

In [85]:
games_region_jp= df.sort_values('jp_sales', ascending=False)
games_region_jp[['name', 'jp_sales']].head(1)

,name,jp_sales
4,Pokemon Red/Pokemon Blue,10.22


Игра "Pokemon Red/Pokemon Blue" занимает лидирующую позицию по продажам в Японии(JP).

##Задача 4. Выпуск игр по годам
Подсчитайте количество игр по годам и выведите таблицу.
Определите год с наибольшим количеством релизов.

In [86]:
df['year_of_release'].isna().sum()    #кол-во пропусков по годам релиза

df['year_of_release'].isna().mean() * 100    #доля пропусков

np.float64(1.609332934489979)

In [ ]:
1.6 % - это небольшой процент пропусков.

In [87]:
df['year_of_release'].value_counts(sort=True).reset_index().head(7)  #пропуски не учитываются по годам


,year_of_release,count
0,2008,1427
1,2009,1426
2,2010,1255
3,2007,1197
4,2011,1136
5,2006,1006
6,2005,939


In [88]:
df['year_of_release'].value_counts(dropna=False).reset_index().head(7)   #учитываются пропуски по годам

,year_of_release,count
0,2008,1427
1,2009,1426
2,2010,1255
3,2007,1197
4,2011,1136
5,2006,1006
6,2005,939


При любом подходе год с наибольшим количеством релизов - 2008г.

##Задача 5. Анализ платформ
Для отфильтрованного датасета выведите таблицу: платформа — суммарные продажи — количество игр — медианные продажи на игру. Определите топ-5 платформ.

In [89]:
top_5_platform = df.groupby('platform').agg(total_sales=('total_sales', 'sum'), count_games=('platform', 'size'), median_sales=('total_sales', 'median')).reset_index()
top_5_platform.sort_values('total_sales', ascending=False).head(5)


,platform,total_sales,count_games,median_sales
16,PS2,1255.77,2161,0.23
28,X360,971.42,1262,0.28
17,PS3,939.65,1331,0.27
26,Wii,907.51,1320,0.19
4,DS,806.12,2151,0.11


#Вывод:
В результате определены топ-5 платформ: PS2, X360, PS3, Wii, DS.

##Задача 6. Анализ жанров
Для каждого жанра рассчитайте суммарные и медианные продажи, количество игр. Выведите таблицу по убыванию суммарных продаж. Определите самый прибыльный жанр и жанр с наибольшими медианными продажами.

In [90]:
genre_new = df.groupby('genre').agg(total_sales=('total_sales', 'sum'), count_games= ('genre', 'size'), median_sales=('total_sales', 'median')).reset_index()
genre_new.sort_values('total_sales', ascending=False). head(5)


,genre,total_sales,count_games,median_sales
0,Action,1744.17,3369,0.19
10,Sports,1331.27,2348,0.22
8,Shooter,1052.45,1323,0.24
7,Role-Playing,934.56,1498,0.18
4,Platform,827.77,888,0.27


Вывод:
Самый прибыльный жанр - это "Action", cуммарные продажи составляют - 1744.
17 млн.копий, медианные продажи - 0.19. Два пропуска по столбцу "genre" не влияют на сумму продаж.


##Задача 7. Региональные предпочтения: жанры и рейтинг
Для каждого региона выведите топ-5 жанров по продажам.
Рассчитайте суммарные продажи по рейтингу ESRB в каждом регионе. Какой рейтинг лидирует?

In [91]:
genre_na_top5 = df.groupby('genre').agg(na_total_sales = ('na_sales', 'sum')).reset_index()

genre_na_top5.sort_values('na_total_sales', ascending=False).head(5)

,genre,na_total_sales
0,Action,879.01
10,Sports,684.43
8,Shooter,592.24
4,Platform,445.50
3,Misc,407.27


#Вывод
Выведены топ-5 продаж по жанрам для Севeрной Америки.

In [92]:
genre_eu_top5 = df.groupby('genre').agg(eu_total_sales = ('eu_sales', 'sum')).reset_index()

genre_eu_top5.sort_values(['eu_total_sales'], ascending=[False]).head(5)


,genre,eu_total_sales
0,Action,519.13
10,Sports,376.79
8,Shooter,317.34
6,Racing,236.51
3,Misc,212.74


#Вывод
Выведены топ-5 продаж по жанрам для Европы.

In [93]:
genre_jp_top5 = df.groupby('genre').agg(jp_total_sales = ('jp_sales', 'sum')).reset_index()
genre_jp_top5.sort_values('jp_total_sales', ascending=False).head(5)

,genre,jp_total_sales
7,Role-Playing,355.41
0,Action,161.43
10,Sports,135.54
4,Platform,130.83
3,Misc,108.11


#Вывод
Выведены топ-5 продаж по жанрам для Японии. В отличие от Северной Америки и Европы лидирует жанр Role-Playing.

#Рейтинг ESRB

1. Необходимо вывести категории рейтинга.

In [108]:
rating_summary = df['rating'].value_counts(dropna=False).reset_index()
rating_summary.columns = ['rating', 'count']
print(rating_summary)

  rating  count
0    NaN   6766
1      E   3990
2      T   2961
3      M   1563
4   E10+   1420
5     EC      8
6    K-A      3
7     RP      3
8     AO      1


2. В столбце "rating" 6766 пропусков.
Необходимо определить пропуски среди игр по рейтингу в разрезе регионов: Северной Америки, Европы и Японии.


In [111]:
# Пропуски рейтинга среди игр, продававшихся в Северной Америке
na_mask = df['na_sales'].notna() & (df['na_sales'] > 0)
missing_rating_in_na = df.loc[na_mask, 'rating'].isna().sum()
percent_na = (missing_rating_in_na / na_mask.sum() * 100) if na_mask.sum() > 0 else 0

# Пропуски рейтинга среди игр, продававшихся в Европе
eu_mask = df['eu_sales'].notna() & (df['eu_sales'] > 0)
missing_rating_in_eu = df.loc[eu_mask, 'rating'].isna().sum()
percent_eu = (missing_rating_in_eu / eu_mask.sum() * 100) if eu_mask.sum() > 0 else 0

# Пропуски рейтинга среди игр, продававшихся в Японии
jp_mask = df['jp_sales'].notna() & (df['jp_sales'] > 0)
missing_rating_in_jp = df.loc[jp_mask, 'rating'].isna().sum()
percent_jp = (missing_rating_in_jp / jp_mask.sum() * 100) if jp_mask.sum() > 0 else 0

print("Пропусков rating среди игр с продажами:")
print(f"NA: {missing_rating_in_na} ({percent_na:.2f} %)")
print(f"EU: {missing_rating_in_eu} ({percent_eu:.2f} %)")
print(f"JP: {missing_rating_in_jp} ({percent_jp:.2f} %)")


Пропусков rating среди игр с продажами:
NA: 3034 (24.85 %)
EU: 3029 (27.93 %)
JP: 3998 (64.47 %)


Доля пропусков в рейтинге среди игр: в Северной Америке - 24,9%, в Европе - 28%, в Японии - 64,5%.

In [100]:
leader_na = df.groupby('rating').agg(na_total=('na_sales', 'sum')).reset_index()
leader_na.sort_values('na_total', ascending=False).head(2)


,rating,na_total
1,E,1292.99
7,T,759.75


In [101]:
leader_eu = df.groupby('rating').agg(eu_total=('eu_sales', 'sum')).reset_index()
leader_eu.sort_values('eu_total', ascending=False).head(2)

,rating,eu_total
1,E,710.25
5,M,483.97


In [105]:
leader_jp = df.groupby('rating').agg(jp_total=('jp_sales', 'sum')).reset_index()
leader_jp.sort_values('jp_total', ascending=False).head(2)

,rating,jp_total
1,E,198.11
7,T,151.40


# Вывод
В современной системе рейтингов ESRB категории “K‑A” и “EC” не используются: они были упразднены. По содержанию игры из этих категорий соответствуют рейтингу “E”. В текущем датасете их доля незначительна, поэтому они не оказывают существенного влияния на итоговые метрики.

Если рассматривать отдельно суммарные продажи в разрезе каждого региона, то в Северной Америке и Европе лидирует рейтинг "Е".

В Японии по данным рейтинга много пропусков, что составляет 64,5% продаж среди игр для данного региона. Соответственно полученные данные по рейтингу не отражают реальную картину для данного региона, т.е. для Японии вывод не презентативен.